cpp/src/join/nested_loop_join.cuh

```diff
@@ -28,167 +29,153 @@
+                             join_kind JoinKind,
+                             ast::expression binary_pred,
+                             null_equality compare_nulls,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
 {
-  const size_type left_num_rows{left.num_rows()};
-  const size_type right_num_rows{right.num_rows()};
-
-  if (right_num_rows == 0) {
-    // If the right table is empty, we know exactly how large the output
-    // will be for the different types of joins and can return immediately
+  // We can immediately filter out cases where the right table is empty. In
+  // some cases, we return all the rows of the left table with a corresponding
+  // null index for the right table; in others, we return an empty output.
+  if (right.num_rows() == 0) {
     switch (JoinKind) {
-      // Inner join with an empty table will have no output
-      case join_kind::INNER_JOIN: return 0;
-
-      // Left join with an empty table will have an output of NULL rows
-      // equal to the number of rows in the left table
-      case join_kind::LEFT_JOIN: return left_num_rows;
-
-      default: CUDF_FAIL("Unsupported join type");
+      // Left, left anti, and full (which are effectively left because we are
+      // guaranteed that left has more rows than right) all return a all the
+      // row indices from left with a corresponding NULL from the right.
+      case join_kind::LEFT_JOIN:
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::FULL_JOIN: return get_trivial_left_join_indices(left, stream);
+      // Inner and left semi joins return empty output because no matches can exist.
+      case join_kind::INNER_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                              std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
     }
   }
 
-  // Allocate storage for the counter used to get the size of the join output
-  size_type h_size_estimate{0};
-  rmm::device_scalar<size_type> size_estimate(0, stream);
+  // Prepare output column. Whether or not the output column is nullable is
+  // determined by whether any of the columns in the input table are nullable.
+  // If none of the input columns actually contain nulls, we can still use the
+  // non-nullable version of the expression evaluation code path for
+  // performance, so we capture that information as well.
+  auto const nullable =
+    std::any_of(left.begin(), left.end(), [](column_view c) { return c.nullable(); }) ||
+    std::any_of(right.begin(), right.end(), [](column_view c) { return c.nullable(); });
+  auto const has_nulls =
+    nullable &&
+    (std::any_of(
+       left.begin(), left.end(), [](column_view c) { return c.nullable() && c.has_nulls(); }) ||
+     std::any_of(
+       right.begin(), right.end(), [](column_view c) { return c.nullable() && c.has_nulls(); }));
```

Comment:
I think there are already table apis for `has_nulls` and `nullable` that already do this.

Reply:
You're right that there's an API for `has_nulls`, but not for `nullable`. I can add the `nullable` one, but perhaps that's out of scope for this PR?

Reply:
Discussed offline, we'll go ahead and add `table.nullable` in this PR.